File: modules/commands/src/main/java/org/polyfrost/oneconfig/api/commands/v1/CommandManager.java
/*
 * This file is part of OneConfig.
 * OneConfig - Next Generation Config Library for Minecraft: Java Edition
 * Copyright (C) 2021~2024 Polyfrost.
 *   <https://polyfrost.org> <https://github.com/Polyfrost/>
 *
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 *   OneConfig is licensed under the terms of version 3 of the GNU Lesser
 * General Public License as published by the Free Software Foundation, AND
 * under the Additional Terms Applicable to OneConfig, as published by Polyfrost,
 * either version 1.0 of the Additional Terms, or (at your option) any later
 * version.
 *
 *   This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public
 * License.  If not, see <https://www.gnu.org/licenses/>. You should
 * have also received a copy of the Additional Terms Applicable
 * to OneConfig, as published by Polyfrost. If not, see
 * <https://polyfrost.org/legal/oneconfig/additional-terms>
 */

package org.polyfrost.oneconfig.api.commands.v1;

import com.mojang.brigadier.arguments.*;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import com.mojang.brigadier.tree.LiteralCommandNode;
import dev.deftu.omnicore.api.client.commands.OmniClientCommandSource;
import dev.deftu.omnicore.api.client.commands.OmniClientCommands;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.polyfrost.oneconfig.api.commands.v1.factories.CommandFactory;
import org.polyfrost.oneconfig.api.commands.v1.factories.annotated.AnnotationCommandFactory;
import org.polyfrost.oneconfig.api.commands.v1.factories.annotated.Command;
import org.polyfrost.oneconfig.utils.v1.WrappingUtils;

import java.util.HashSet;
import java.util.IdentityHashMap;
import java.util.Map;
import java.util.Set;

/**
 * Handles the registration of OneConfig commands.
 *
 * @see Command
 */
public class CommandManager {

    private static final Logger LOGGER = LogManager.getLogger("OneConfig/CommandManger");

    /**
     * The singleton instance of the command manager.
     */
    public static final CommandManager INSTANCE = new CommandManager();
    private final Set<CommandFactory> factories = new HashSet<>();
    private final Map<Class<?>, ArgumentType<?>> argumentTypes = new IdentityHashMap<>();


    private CommandManager() {
        argumentTypes.put(int.class, IntegerArgumentType.integer());
        argumentTypes.put(String.class, StringArgumentType.string());
        argumentTypes.put(String[].class, StringArgumentType.greedyString());
        argumentTypes.put(boolean.class, BoolArgumentType.bool());
        argumentTypes.put(float.class, FloatArgumentType.floatArg());
        argumentTypes.put(double.class, DoubleArgumentType.doubleArg());
        argumentTypes.put(long.class, LongArgumentType.longArg());
        registerFactory(new AnnotationCommandFactory());
    }

    public static void register(LiteralArgumentBuilder<OmniClientCommandSource> builder) {
        register(builder.build());
    }

    public static void register(LiteralCommandNode<OmniClientCommandSource> node) {
        OmniClientCommands.register(node);
    }

    public static boolean register(Object obj) {
        LiteralCommandNode<OmniClientCommandSource>[] nodes = INSTANCE.create(obj);
        if (nodes == null) {
            return false;
        }

        for (LiteralCommandNode<OmniClientCommandSource> node : nodes) {
            if (node == null) {
                LOGGER.warn("Command node for object {} is null, skipping registration", obj);
                continue;
            }

            OmniClientCommands.register(node);
//            LOGGER.info("Registered command node for object {}: {}", obj, node.getName());
        }

        return true;
    }

    public LiteralCommandNode<OmniClientCommandSource>[] create(Object obj) {
        for (CommandFactory factory : factories) {
            try {
                LiteralCommandNode<OmniClientCommandSource>[] nodes = factory.create(obj);
                if (nodes != null) {
                    return nodes;
                }
            } catch (Exception e) {
                LOGGER.error("Command factory {} failed with an exception trying to create commands for object {}", factory, obj, e);
            }
        }

        return null;
    }

    public static LiteralArgumentBuilder<OmniClientCommandSource> literal(String name) {
        return OmniClientCommands.literal(name);
    }

    public static <T> RequiredArgumentBuilder<OmniClientCommandSource, T> argument(String name, ArgumentType<T> type) {
        if (type == null) {
            throw new NullPointerException("Can't get argument type for argument '" + name + "' as type is null");
        }

        return OmniClientCommands.argument(name, type);
    }

    /**
     * Register a factory which can be used to create commands from objects in the {@link #create(Object)} method.
     */
    public void registerFactory(CommandFactory factory) {
        factories.add(factory);
    }

    public <T> void registerArgumentType(Class<T> type, ArgumentType<T> argumentType) {
        argumentTypes.put(type, argumentType);
    }

    @SuppressWarnings("unchecked")
    public static <T> ArgumentType<T> getArgumentType(Class<T> type) {
        return (ArgumentType<T>) INSTANCE.argumentTypes.get(WrappingUtils.getUnwrapped(type));
    }

}

File: minecraft/src/main/kotlin/org/polyfrost/oneconfig/api/commands/v1/CommandManager.kt
package org.polyfrost.oneconfig.api.commands.v1

import com.mojang.brigadier.arguments.ArgumentType
import com.mojang.brigadier.arguments.BoolArgumentType
import com.mojang.brigadier.arguments.DoubleArgumentType
import com.mojang.brigadier.arguments.FloatArgumentType
import com.mojang.brigadier.arguments.IntegerArgumentType
import com.mojang.brigadier.arguments.LongArgumentType
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.builder.RequiredArgumentBuilder
import com.mojang.brigadier.tree.LiteralCommandNode
import net.fabricmc.fabric.api.client.command.v2.ClientCommandRegistrationCallback
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.polyfrost.oneconfig.api.commands.v1.factories.CommandFactory
import org.polyfrost.oneconfig.api.commands.v1.factories.annotated.AnnotationCommandFactory
import org.polyfrost.oneconfig.utils.v1.WrappingUtils
import java.util.*


/**
 * Handles the registration of OneConfig commands
 *
 * @see Command
 */
typealias ClientCommandSource = FabricClientCommandSource

object CommandManager {
    private val factories: MutableSet<CommandFactory> = HashSet<CommandFactory>()
    private val argumentTypes: MutableMap<Class<*>, ArgumentType<*>> = IdentityHashMap<Class<*>, ArgumentType<*>>()


    init {
        argumentTypes[Int::class.javaPrimitiveType!!] = IntegerArgumentType.integer()
        argumentTypes[String::class.java] = StringArgumentType.string()
        argumentTypes[Array<String>::class.java] = StringArgumentType.greedyString()
        argumentTypes[Boolean::class.javaPrimitiveType!!] = BoolArgumentType.bool()
        argumentTypes[Float::class.javaPrimitiveType!!] = FloatArgumentType.floatArg()
        argumentTypes[Double::class.javaPrimitiveType!!] = DoubleArgumentType.doubleArg()
        argumentTypes[Long::class.javaPrimitiveType!!] = LongArgumentType.longArg()
        registerFactory(AnnotationCommandFactory())

        ClientCommandRegistrationCallback.EVENT.register { dispatcher, context ->
            commands.forEach { dispatcher.root.addChild(it) }
        }
    }

    fun create(obj: Any): Array<LiteralCommandNode<ClientCommandSource>?>? {
        for (factory in factories) {
            try {
                val nodes = factory.create(obj)
                if (nodes != null) {
                    return nodes
                }
            } catch (e: Exception) {
                LOGGER.error(
                    "Command factory {} failed with an exception trying to create commands for object {}",
                    factory,
                    obj,
                    e
                )
            }
        }

        return null
    }

    /**
     * Register a factory which can be used to create commands from objects in the [.create] method
     */
    fun registerFactory(factory: CommandFactory) {
        factories.add(factory)
    }

    fun <T> registerArgumentType(type: Class<T>, argumentType: ArgumentType<T>) {
        argumentTypes[type] = argumentType
    }

    private val commands = mutableListOf<LiteralCommandNode<ClientCommandSource>>()

    private val LOGGER: Logger = LogManager.getLogger("OneConfig/CommandManger")


    fun register(builder: LiteralArgumentBuilder<ClientCommandSource>) {
        register(builder.build())
    }

    fun register(node: LiteralCommandNode<ClientCommandSource>) {
        commands.add(node)
    }

    @JvmStatic
    fun register(obj: Any): Boolean {
        val nodes: Array<LiteralCommandNode<ClientCommandSource>?> = create(obj) ?: return false

        for (node in nodes) {
            if (node == null) {
                LOGGER.warn("Command node for object {} is null, skipping registration", obj)
                continue
            }

            commands.add(node)
        }

        return true
    }

    @JvmStatic
    fun literal(name: String): LiteralArgumentBuilder<ClientCommandSource> {
        return LiteralArgumentBuilder.literal<ClientCommandSource>(name)
    }

    fun <T> argument(name: String, type: ArgumentType<T>): RequiredArgumentBuilder<ClientCommandSource, T> {
        return RequiredArgumentBuilder.argument<ClientCommandSource, T>(name, type)
    }

    fun <T> getArgumentType(type: Class<T>): ArgumentType<T>? {
        @Suppress("UNCHECKED_CAST")
        return argumentTypes[WrappingUtils.getUnwrapped(type)] as ArgumentType<T>?
    }

}